### PR TITLE
[2162] Unable to perform two chained direct-edit

### LINKED ISSUE #2162: Unable to perform two chained direct-edit
### Steps to reproduce:

- Create a papaya instance
- Create a papaya diagram rendered with react flow
- Create a node
- Trigger the direct-edit on the created node, and update its name
- Try to trigger the direct-edit again using "f2" or starting to type

### Actual behavior

The second direct-edit is not triggered

### Expected behavior

The text field used to edit the node label should be displayed

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2162

# Pull request template

## General purpose
What is the main goal of this pull request?
- [x] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/direct-edit/DiagramDirectEditInput.tsx [hunk 1]
```diff
@@ -88,6 +94,15 @@ export const DiagramDirectEditInput = ({ labelId, editingKey, onClose, transform
 
   const [renameElement, { data: editLabelData, error: editLabelError }] = useMutation(editLabelMutationOp);
 
+  useEffect(() => {
+    return () => {
+      let elementToFocus: HTMLDivElement | null = (parentElementRef.current?.parentNode as HTMLDivElement) ?? null;
+      if (elementToFocus) {
+        elementToFocus.focus();
+      }
+    };
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/direct-edit/DiagramDirectEditInput.tsx]: You need a ref to the parent element to force the focus on it.
You may want to support edges too; thus, `HTMLDivElement` may not be appropriate here.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/direct-edit/DiagramDirectEditInput.tsx [hunk 2]
```diff
@@ -88,6 +94,15 @@ export const DiagramDirectEditInput = ({ labelId, editingKey, onClose, transform
 
   const [renameElement, { data: editLabelData, error: editLabelError }] = useMutation(editLabelMutationOp);
 
+  useEffect(() => {
+    return () => {
+      let elementToFocus: HTMLDivElement | null = (parentElementRef.current?.parentNode as HTMLDivElement) ?? null;
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/direct-edit/DiagramDirectEditInput.tsx]: Why do you try to focus on parentNode while `parentElementRef.current` seems to reference the parent?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/edge/MultiLabelEdge.tsx [hunk 3]
```diff
@@ -61,7 +61,7 @@ export const MultiLabelEdge = memo(
     });
 
     const { beginLabel, endLabel, label, faded } = data || {};
-
+    const elementRef = useRef<HTMLDivElement | null>(null);
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/edge/MultiLabelEdge.tsx]: You may need to put a ref on edge too

**Review by gcoutable — DISMISSED** [review]

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
If I understand things correctly, when the direct edit is triggered using the pen icon in the palette then the focus is lost (since it moves to the tool and then to the body of the document) and after that we can't trigger a new direct edit because the keydown listener on the diagram is not triggered.

I don't see why you need to give a ref from everywhere the label component is used to the very bottom of the label component in order to focus the relevant element. Couldn't you simply ask for the `diagramElementId` on Label in order to know the parent of the label element (since it's not the parent in the DOM) and then use something like this:

```
document.querySelector(`[data-id="${diagramElementId}"]`).focus();
```

**Comment by sbegaudeau** [comment]
Since we need to move forward with that, I'll take it from here and ensure that it will be merged.

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**